src/rest_api/api/search/superstructure_search.rs

```diff
@@ -0,0 +1,177 @@
+use crate::rest_api::api::{GetStructureSearchResponse, StructureResponseError};
+use crate::search::scaffold_search::{scaffold_search, PARSED_SCAFFOLDS};
+use crate::search::{
+    compound_processing::{get_cpd_properties, get_tautomers},
+    superstructure_search::superstructure_search,
+    {aggregate_search_hits, prepare_query_structure},
+};
+use poem_openapi::payload::Json;
+
+pub fn v1_index_search_superstructure(
+    index_manager: &IndexManager,
+    index: String,
+    smiles: String,
+    result_limit: usize,
+    tautomer_limit: usize,
+    extra_query: &str,
+    use_scaffolds: bool,
+) -> GetStructureSearchResponse {
+    let index = match index_manager.open(&index) {
+        Ok(index) => index,
+        Err(e) => {
+            return GetStructureSearchResponse::Err(Json(StructureResponseError {
+                error: e.to_string(),
+            }))
+        }
+    };
+
+    let reader = index.reader();
+    let reader = match reader {
+        Ok(reader) => reader,
+        Err(e) => {
+            return GetStructureSearchResponse::Err(Json(StructureResponseError {
+                error: e.to_string(),
+            }))
+        }
+    };
+
+    let searcher = reader.searcher();
+
+    let query_attributes = prepare_query_structure(&smiles);
+
+    let query_attributes = match query_attributes {
+        Ok(query_attributes) => query_attributes,
+        Err(e) => {
+            return GetStructureSearchResponse::Err(Json(StructureResponseError {
+                error: e.to_string(),
+            }))
+        }
+    };
+
+    let (query_canon_taut, fingerprint, descriptors) = query_attributes;
+
+    let scaffolds = if use_scaffolds {
+        Some(&PARSED_SCAFFOLDS)
+    } else {
+        None
+    };
+
+    let matching_scaffolds = match scaffolds {
+        Some(scaffolds) => {
+            let scaffold_matches = scaffold_search(&query_canon_taut, scaffolds);
+            match scaffold_matches {
+                Ok(scaffold_matches) => Some(scaffold_matches),
+                Err(e) => {
+                    return GetStructureSearchResponse::Err(Json(StructureResponseError {
+                        error: e.to_string(),
+                    }))
+                }
+            }
+        }
+        None => None,
+    };
+
+    let results = superstructure_search(
+        &searcher,
+        &query_canon_taut,
+        &matching_scaffolds,
+        fingerprint.0.as_bitslice(),
+        &descriptors,
+        result_limit,
+        extra_query,
+    );
+
+    let mut results = match results {
+        Ok(results) => results,
+        Err(e) => {
+            return GetStructureSearchResponse::Err(Json(StructureResponseError {
+                error: e.to_string(),
+            }))
+        }
+    };
+
+    let mut used_tautomers = false;
+    let mut num_tauts_used = 0;
+    if !results.is_empty() {
+        num_tauts_used = 1;
+    }
+
+    if results.len() < result_limit {
```

Comment:
So I reduced the size a bit mainly by simplifying the scaffold assignment code blocks.

The only thing about the `Iter` chain idea is that we don't necessary want to execute every loop. After each loop, we count the results and see if it's worth trying another tautomer. This could be either because we need more results or because something funky happened when we were testing the previous tautomer (thus the `Err(_) => continue`).

Reply:
`Iter` chain methods have been implemented :)